XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,540 @@
+  asyncHook.remove();
+}, 5000);
+```
+
+
+### Hook Callbacks
+
+Key events in the lifetime of asynchronous events have been categorized into
+four areas. On instantiation, before/after the callback is called and when the
+instance is destructed. For cases where resources are reused, instantiation and
+destructor calls are emulated.
+
+
+#### `init(id, type, parentId, handle)`
+
+* `id` {Number}
+* `type` {String}
+* `parentId` {Number}
+* `handle` {Object}
+
+Called when a class is constructed that has the possibility to trigger an
+asynchronous event. This does not mean the instance will trigger a
+`before()`/`after()` event before `destroy()` is called. Only that the
+possibility exists.
+
+This behavior can be observed by doing something like opening a resource then
+closing it before the resource can be used. The following snippet demonstrates
+this.
+
+```js
+require('net').createServer().listen(8080, function() { this.close() });
+```
+
+Every instance is assigned a unique id. Taking advantage of the fraction space
+in the 64-bit IEEE 754 that ECMAScript defines as the **Number** type, the
+number of id's that can be assigned are `2^53 - 1` (also defined as
+`Number.MAX_SAFE_INTEGER`). At this size node can assign a new id every 100
+nanoseconds and not run out for over 28 years. Because of this circumstance it
+is not deemed necessary to use an alternative approach that would account for
+the id to wrap around.
+
+The `type` is a String that represents the type of handle that caused `init()`
+to fire. Generally it will be the name of the handle's constructor. Some
+examples include `TCP`, `GetAddrInfo` and `HTTPParser`. Users will be able to
+define their own `type` when using the public API.
+
+`parentId` is the unique id of either the async resource at the top of the JS
+stack when `init()` was called, or the originator of the new resource. For
+example, when a connection is made to a TCP server there is no JS stack
+available when the connection's constructor is executed. Meaning there would be
+no `parentId` available on the JS stack. Instead the TCP server's unique id is
+manually passed to the client's constructor and propagated to the `init()`'s
+`parentId`.
+
+The constructed `handle` is passed to `init()`. The structure of any object
+passed to `init()` has no guarantee of stability, even through patch updates.
+It's meant to be used for debugging the application when additional, and
+specific, information is needed. Make sure not to hold a reference to this
+indefinitely or else the GC won't be able to collect it after node has removed
+its own reference to the handle.
+
+
+#### `before(id)`
+
+* `id` {Number}
+
+Called just before the return callback is called after completing an
+asynchronous request. Or called on handles with events such as receiving a new
+connection. For requests, such as `fs.open()`, this should be called exactly
+once. For handles, such as a TCP server, this may be called 0-N times.
+
+
+#### `after(id)`
```

Comment:
Sounds reasonable. It also doesn't fit that well into the `events.emitAfter()` embedder API (but obviously not impossible).